hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -2126,6 +2127,92 @@ private class Emit(
           }
         }
 
+      case x@NDArrayConcat(nds, axis) =>
+        val inputType = coerce[PArray](nds.pType2)
+        val inputNDType = coerce[PNDArray](inputType.elementType)
+
+        val ndType = coerce[PNDArray](x.pType2)
+        val codeNDs = emit(nds, env, er, None)
+
+        val inputArray = mb.newField[Long]
+        val n = mb.newField[Int]
+        val i = mb.newField[Int]
+
+        val loadAndValidateArray = Code(
+          inputArray := codeNDs.value[Long],
+          n := inputType.loadLength(inputArray),
+          (n < 1).orEmpty(Code._fatal("need at least 1 array to concatenate")))
+
+        val (missingSetup: Code[Unit @unchecked], missing: Code[Boolean @unchecked], setupShape: Code[Unit @unchecked]) = (inputType.required, inputNDType.required) match {
+          case (true, true) => (Code._empty, const(false), Code(
+            codeNDs.setup,
+            codeNDs.m.orEmpty(Code._fatal("required NDArray can't be missing")),
+            loadAndValidateArray))
+          case (false, true) => (codeNDs.setup, codeNDs.m, loadAndValidateArray)
+          case _ =>
+            val m = mb.newField[Boolean]
+            val setup = Code(
+              codeNDs.setup,
+              m := codeNDs.m,
+              loadAndValidateArray,
+              i := 0,
+              Code.whileLoop(i < n,
+                m := m | inputType.isElementMissing(inputArray, i),
+                i := i + 1))
+            (setup, m.load(), Code._empty)
+        }
+
+        val localDim = mb.newField[Long]
+        val outputShape = Array.tabulate(ndType.nDims) { idx =>
+          if (idx == axis) {
```

Comment:
Your choice, but this `if` might be easier as 

```
Code(
    localDim := inputNDType.dimensionLength(inputType.loadElement(inputArray, 0), idx),
    Code.forLoop(i := 1, i < n, i := i + 1, {
        if (idx == axis) {
            localDim := localDim + inputNDType.dimensionLength(inputType.loadElement(inputArray, i), idx)
        }
        else {
            inputNDType.dimensionLength(inputType.loadElement(inputArray, i), idx).cne(localDim).orEmpty(Code._fatal("mismatched dims"))
        }
    })
)
```

Saves you the trouble of setting up the loop twice.